protostar/starknet/new_arch/cheatable_cairo_cached_state.py

```diff
@@ -0,0 +1,176 @@
+from pathlib import Path
+from typing import Dict, List
+
+from starkware.starknet.business_logic.state.state import (
+    ContractClassCache,
+    CachedState,
+)
+from starkware.starknet.public.abi import AbiType
+from starkware.starknet.business_logic.state.state_api_objects import BlockInfo
+from starkware.starknet.business_logic.state.state_api import (
+    StateReader,
+)
+from typing_extensions import Self
+
+from protostar.starknet.address import Address
+from protostar.starknet.new_arch.cheaters import BlockInfoCairoCheater
+from protostar.starknet.types import ClassHashType, SelectorType
+from protostar.starknet.data_transformer import CairoOrPythonData
+
+
+# pylint: disable=too-many-instance-attributes
+class CheatableCairoCachedState(CachedState):
```

Comment:
> the name does not reflect that

The namings is consistent with all of the components which are relevant to cairo testing flow. It was called `CheatableCachedState` previously, so no major changes introduced in naming. 

> I would create a component called `protostar.cheatable_starknet_fixture`

It's not really a fixture as well, since it's extension of `state`. 

> Similarly, I'd create a new component `protostar.cairo_testing` instead of `protostar.testing.new_arch.`

The idea i had in putting those new counterparts in `new_arch`, is that when we remove old flow completely, we can just copy things over one catalogue higher and ditch the `new_arch` prefix. I think creating completely new structure just for sake of refactoring at this point may result in confusion and will make the rewrite less workable.

Reply:
> The namings is consistent with all of the components which are relevant to cairo testing flow. It was called CheatableCachedState previously, so no major changes introduced in naming.

The naming is consistent, although it's confusing. `CheatableCairoCachedState` on its own could be interpreted in two invalid ways:
- `CheatableCairo` `CachedState` — Protostar cheats on StarkNet, not Cairo.
- `Cheatable` `CairoCachedState` — It's StarkNet's state, not Cairo's.

`Cairo` prefix makes sense within `protostar.testing` component, but it doesn't in `protostar.starknet`. This could in a different PR.

> It's not really a fixture as well, since it's extension of state.

In this case, I meant renaming/relocating the directory `new_arch`, not the class name.

> The idea i had in putting those new counterparts in new_arch, is that when we remove old flow completely, we can just copy things over one catalogue higher and ditch the new_arch prefix. I think creating completely new structure just for sake of refactoring at this point may result in confusion and will make the rewrite less workable.

Ditching `new_arch` will also create a confusion. Refactoring often introduces confusion, but it's temporary. This PR changes the structure anyway.

---

After looking at this PR again, I'd move these components in the following way:
`protostar.starknet.new_arch` -> `protostar.cheatable_starknet`
`protostar.testing.new_arch` -> `protostar.cheatable_starknet_fixture`

`protostar.starknet` and `protostar.testing` are already quite large, and this is a good opportunity to move stuff out of them.

Reply:
> After looking at this PR again, I'd move these components in the following way:
`protostar.starknet.new_arch` -> `protostar.cheatable_starknet`
`protostar.testing.new_arch` -> `protostar.cheatable_starknet_fixture`

Agree with the first, but the second one includes execution environment, test runner, etc. 
I would rather go with the first proposition and name it `cairo_testing`

Reply:
As for the name of `CheatableCairoCachedState` i would gladly accept another name if you have ideas. `StarknetCheatableCachedState` is not acceptable because `StarknetState` is not an instance of `CachedState` - those are two different concepts

Reply:
> I would rather go with the first proposition and name it cairo_testing

Ok, let it be `protostar.cairo_testing`. I suggested `protostar.cheatable_starknet_fixture` because `protostar.cairo_testing` should be unaware of StarkNet.

> As for the name of CheatableCairoCachedState i would gladly accept another name if you have ideas. StarknetCheatableCachedState is not acceptable because StarknetState is not an instance of CachedState - those are two different concepts

I'd keep `CheatableCachedState` as long as the name is unique within the component. Old classes are going to be removed anyway. If the name must be globally unique, I'd prefix old `Cheatable` classes with `Deprecated` e.g. `DeprecatedCheatableCachedState`.

Reply:
Done as discussed